Sources/Testing/Test+Discovery.swift

```diff
@@ -80,32 +80,23 @@ extension Test {
   ///
   /// - Parameters:
   ///   - tests: A dictionary of tests to amend.
-  ///
-  /// - Returns: The number of key-value pairs added to `tests`.
-  @discardableResult private static func _synthesizeSuiteTypes(into tests: inout [ID: Self]) -> Int {
-    let originalCount = tests.count
-
-    // Find any instances of Test in the input that are *not* suites. We'll be
-    // checking the containing types of each one.
-    for test in tests.values where !test.isSuite {
-      guard let suiteTypeInfo = test.containingTypeInfo else {
-        continue
-      }
-      let suiteID = ID(typeInfo: suiteTypeInfo)
-      if tests[suiteID] == nil {
-        tests[suiteID] = Test(traits: [], sourceLocation: test.sourceLocation, containingTypeInfo: suiteTypeInfo, isSynthesized: true)
-
-        // Also synthesize any ancestral suites that don't have tests.
-        for ancestralSuiteTypeInfo in suiteTypeInfo.allContainingTypeInfo {
-          let ancestralSuiteID = ID(typeInfo: ancestralSuiteTypeInfo)
-          if tests[ancestralSuiteID] == nil {
-            tests[ancestralSuiteID] = Test(traits: [], sourceLocation: test.sourceLocation, containingTypeInfo: ancestralSuiteTypeInfo, isSynthesized: true)
-          }
+  private static func _synthesizeSuiteTypes(into tests: inout [ID: Self]) {
+    // We won't have an accurate source location for synthesized suites, so use
+    // a "runtime inferred" sentinel value instead.
+    //
+    // FIXME: Consider either changing `Test.sourceLocation` to Optional (an
+    // API-breaking change) or modifying Runner.Plan so that it doesn't require
+    // an actual Test instance to represent synthesized parent suites.
+    lazy var inferredSourceLocation = SourceLocation(fileID: "<runtime-inferred>", filePath: "<runtime-inferred>", line: 0, column: 0)
+
+    for test in tests.values {
```

Comment:
What do the performance characteristics of this look like? Do we possibly need more caching to keep the repeated lookups from becoming a bottleneck?

Reply:
I don't think the performance characteristics of this function are changed in a significant way. However, I have never really loved the approach that this method uses, which involves mapping the tests into a dictionary, iterating every test and all of its ancestors, constructing a test ID for each one, and testing whether it's already in the dictionary.

I decided to take a different approach with this PR, which requires broader changes but I think has more benefits overall. In my latest commit, we no longer construct this temporary dictionary, and instead synthesize suites during planning while traversing the test graph. While implementing and debugging this, I also noticed some opportunities to simplify and optimize the test graph by pruning nodes for tests which have been filtered out.

Reply:
The dictionary (or a set) is necessary to avoid duplicate test or suite records which can potentially appear multiple times in the raw data we're walking. Does your new code dedup the data?

Reply:
Yeah, they are de-duplicated by their placement in the test graph (which relies on a dictionary)